File: scripts/check-liver-bounds.mjs
import fs from'node:fs';import {NodeIO}from'@gltf-transform/core';import{ALL_EXTENSIONS}from'@gltf-transform/extensions';import draco from'draco3dgltf';import *as T from'three';import{hepaticPoint,isHepatic}from'../lib/hepatic.js';
const io=new NodeIO().registerExtensions(ALL_EXTENSIONS).registerDependencies({'draco3d.decoder':await draco.createDecoderModule()});
const bones=await io.read('public/models/skeleton-web.glb'),organs=await io.read('public/models/visceral-web.glb');
const ribs={points:[],faces:[]},liver=[];
function extract(n){const m=new T.Matrix4().fromArray(n.getWorldMatrix());return n.getMesh().listPrimitives().map(p=>{const a=p.getAttribute('POSITION').getArray(),positions=[];for(let i=0;i<a.length;i+=3)positions.push(new T.Vector3().fromArray(a,i).applyMatrix4(m));return {positions,indices:Array.from(p.getIndices().getArray())};});}
for(const n of bones.getRoot().listNodes())if(n.getMesh()&&/rib|sternum|vertebra t/i.test(n.getName()))for(const g of extract(n)){const base=ribs.points.length;ribs.points.push(...g.positions.map(v=>v.toArray()));for(let i=0;i<g.indices.length;i+=3)ribs.faces.push(g.indices.slice(i,i+3).map(j=>j+base));}
for(const n of organs.getRoot().listNodes())if(n.getMesh()&&isHepatic(n.getName()))for(const g of extract(n))liver.push({name:n.getName(),indices:g.indices,rest:g.positions.map(p=>hepaticPoint(p,0).toArray()),inhale:g.positions.map(p=>hepaticPoint(p,1).toArray())});
fs.writeFileSync('/tmp/soma-liver-check.json',JSON.stringify({ribs,liver}));

File: scripts/check-liver-bounds.py
import json
from pathlib import Path
from mathutils import Vector
from mathutils.bvhtree import BVHTree
s=json.loads(Path('/tmp/soma-liver-check.json').read_text());rib=BVHTree.FromPolygons(s['ribs']['points'],s['ribs']['faces'],all_triangles=True);gap=1
for m in s['liver']:
 a=[Vector(v) for v in m['rest']];b=[Vector(v) for v in m['inhale']];ix=m['indices'];faces=[ix[i:i+3] for i in range(0,len(ix),3)]
 for step in range(21):
  points=[v.lerp(w,step/20) for v,w in zip(a,b)]
  tree=BVHTree.FromPolygons(points,faces,all_triangles=True)
  overlaps=rib.overlap(tree)
  assert not overlaps,(m['name'],step,len(overlaps))
  for p in points:
   d=rib.find_nearest(p)[3]
   if d<gap:gap=d;worst=(m['name'],step,list(p),list(rib.find_nearest(p)[0]))
print('Closest rib sample',worst)
assert gap>.003,gap
print('Liver + gallbladder: no rib intersections in 21 respiratory phases; minimum vertex clearance (mm)',gap*1000)
